File: lib/src/db/entities/incident.entity.ts
import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn
} from "typeorm";
import { User } from "./user.entity";
import { Project } from "./project.entity";
import { BaseEntity } from "../../common/base/base.entity";
import { IIncident, IProject, Platform, SDK, Trace } from "@traceo/types";

export enum IncidentStatus {
  RESOLVED = "resolved",
  UNRESOLVED = "unresolved",
  IN_PROGRESS = "in_progress"
}

@Entity()
export class Incident extends BaseEntity implements IIncident {
  @PrimaryGeneratedColumn("uuid")
  id?: string;

  @Column({
    type: "varchar",
    nullable: false
  })
  sdk: SDK;

  @Column({
    type: "varchar",
    nullable: false
  })
  status: IncidentStatus;

  @Column({
    type: "varchar",
    nullable: true
  })
  stack: string;

  @Column({
    type: "varchar",
    nullable: false
  })
  name: string;

  @Column({
    type: "varchar",
    nullable: true
  })
  message: string;

  @Column({
    type: "bigint",
    nullable: false,
    name: "last_event_at"
  })
  lastEventAt: number = this.createdAt;

  @Column({
    type: "bigint",
    nullable: false,
    name: "events_count"
  })
  eventsCount: number = 0;

  @ManyToOne(() => Project, {
    onUpdate: "CASCADE",
    onDelete: "CASCADE"
  })
  @JoinColumn({
    name: "project_id"
  })
  project: IProject;

  @ManyToOne(() => User, (user) => user.incidents, {
    onDelete: "SET NULL"
  })
  @JoinColumn({
    name: "assigned_id"
  })
  assigned: User;

  @Column({
    type: "simple-json",
    nullable: true
  })
  platform: Platform;

  @Column({
    type: "simple-json",
    nullable: true
  })
  traces: Array<Trace>;
}

File: server/traceo-common/traceo-common-jpa/src/main/java/org/traceo/common/jpa/entities/IncidentEntity.java
package org.traceo.common.jpa.entities;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.traceo.common.transport.dto.IncidentTraceDto;
import org.traceo.common.transport.enums.IncidentStatus;
import org.traceo.common.transport.enums.TraceoSdk;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Entity
@Table(name = "traceo_incident")
@Getter
@Setter
public class IncidentEntity extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(nullable = false)
    private TraceoSdk sdk;

    @Column(nullable = false)
    private IncidentStatus status;

    @Column(nullable = false)
    private String name;

    private String stack;
    private String message;

    @Column(name = "last_event_at")
    private BigDecimal lastEventAt;

    @Column(name = "events_count")
    private Long eventsCount = 0L;

    @ManyToOne(cascade = {
            CascadeType.REMOVE,
            CascadeType.PERSIST
    })
    @JoinColumn(name = "project_id")
    private ProjectEntity project;

    @ManyToOne()
    @JoinColumn(name = "assigned_id")
    private UserEntity assigned;

    @JdbcTypeCode(value = SqlTypes.JSON)
    private Map<String, String> platform = new HashMap<>();

    @JdbcTypeCode(value = SqlTypes.JSON)
    private List<IncidentTraceDto> traces = new ArrayList<>();
}
